src/mesh.cpp

```diff
@@ -213,88 +215,105 @@ void Mesh::material_volumes(
-            break;
+          // Set birth cell attribute
+          if (p.cell_born() == C_NONE)
+            p.cell_born() = p.lowest_coord().cell;
 
+          // Initialize last cells from current cell
           for (int j = 0; j < p.n_coord(); ++j) {
             p.cell_last(j) = p.coord(j).cell;
           }
           p.n_coord_last() = p.n_coord();
 
-          // Set surface that particle is on and adjust coordinate levels
-          p.surface() = boundary.surface_index;
-          p.n_coord() = boundary.coord_level;
-
-          if (boundary.lattice_translation[0] != 0 ||
-              boundary.lattice_translation[1] != 0 ||
-              boundary.lattice_translation[2] != 0) {
-            // Particle crosses lattice boundary
-
-            cross_lattice(p, boundary);
-          } else {
-            // Particle crosses surface
-            // TODO: off-by-one
-            const auto& surf {model::surfaces[std::abs(p.surface()) - 1].get()};
-            p.cross_surface(*surf);
+          while (true) {
+            // Ray trace from r_start to r_end
+            Position r0 = p.r();
+            double max_distance = bbox.max()[axis] - r0[axis];
+
+            // Find the distance to the nearest boundary
+            BoundaryInfo boundary = distance_to_boundary(p);
+
+            // Advance particle forward
+            double distance = std::min(boundary.distance, max_distance);
+            p.move_distance(distance);
+
+            // Determine what mesh elements were crossed by particle
+            bins.clear();
+            length_fractions.clear();
+            this->bins_crossed(r0, p.r(), p.u(), bins, length_fractions);
+
+            // Add volumes to any mesh elements that were crossed
+            int i_material = p.material();
+            if (i_material != C_NONE) {
+              i_material = model::materials[i_material]->id();
+            }
+            for (int i_bin = 0; i_bin < bins.size(); i_bin++) {
+              int mesh_index = bins[i_bin];
+              double length = distance * length_fractions[i_bin];
+
+              // Add volume to result
+              result.add_volume(mesh_index, i_material, length * d1 * d2);
+            }
+
+            if (distance == max_distance)
+              break;
+
+            for (int j = 0; j < p.n_coord(); ++j) {
+              p.cell_last(j) = p.coord(j).cell;
+            }
+            p.n_coord_last() = p.n_coord();
+
+            // Set surface that particle is on and adjust coordinate levels
+            p.surface() = boundary.surface_index;
+            p.n_coord() = boundary.coord_level;
+
+            if (boundary.lattice_translation[0] != 0 ||
+                boundary.lattice_translation[1] != 0 ||
+                boundary.lattice_translation[2] != 0) {
+              // Particle crosses lattice boundary
+
+              cross_lattice(p, boundary);
+            } else {
+              // Particle crosses surface
+              // TODO: off-by-one
```

Comment:
What's the off-by-one situation you need to handle?

Reply:
It has to do with surface half-spaces; see the discussion here: https://github.com/openmc-dev/openmc/issues/1174#issuecomment-814912993